### PR TITLE
Swept Langmuir | Ion-Saturation Current

### PR DESCRIPTION
**Pull request by rocco8773**
This is part of the Swept Langmuir analysis overhaul, issue #1402 .  This PR introduces `plasmapy.analysis.swept_langmuir.find_ion_saturation_current` for fitting the ion-saturation current region of an IV-curve.  The linear portion of the fit, representing the ion-saturation current, is returned to the user.

The associated notebook will be created in a followup PR.

### CODE REVIEW

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (namurphy) [inline plasmapy/analysis/swept_langmuir/ion_saturation_current.py:55]:
```suggestion
    The current collected by a Langmuir probe reaches ion-saturation
    when the probe is sufficiently biased so the influx of electrons is
    completely repelled, which leads to only the collection of ions.  (For
```
[optional]

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (namurphy) [inline plasmapy/analysis/swept_langmuir/ion_saturation_current.py:74]:
```suggestion
        Defaults to ``"exp_plus_linear"``.
```
One thing I learned somewhat recently is that using [all caps](https://en.wikipedia.org/wiki/All_caps) has generally been found to reduce legibility.

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (namurphy) [inline plasmapy/analysis/swept_langmuir/ion_saturation_current.py:107]: Sooooooo tempted to suggest a unicode character here even though it would be a bad idea!

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (namurphy) [inline plasmapy/analysis/swept_langmuir/ion_saturation_current.py:179]:
```suggestion
            "Both keywords 'current_bound' and 'voltage_bound' are specified, "
```

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (namurphy) [inline plasmapy/analysis/swept_langmuir/ion_saturation_current.py:51]: [optional] I'm wondering if ```:math:`I_{sat}` ``` is used often enough to warrant a reST substitution in `docs/common_links.rst` (possibly with the `\` at the end to help with spacing).

**Review by namurphy — COMMENTED** [review]
It's looking good!  I have only minor/unimportant suggestions, most of which you are free to ignore or adopt as you see fit.  I could probably approve this pretty quickly after it's reviewed by someone who has a much better understanding of Langmuir stuff than I do.

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (rocco8773) [reply plasmapy/analysis/swept_langmuir/ion_saturation_current.py:51]: At this point I don't thinks so, but might I might change my tune as the `swept_langmuir` module expands.

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (rocco8773) [reply plasmapy/analysis/swept_langmuir/ion_saturation_current.py:75]: While I generally agree with the statement about the readability of all caps, I don't think it applies here.  The all caps statement applies to general writings (emails, books, journals, etc.), but we have something a little more specialized here.  One of my biggest issues with documenting functionality is that default values are buried within the documentation, making it rather difficult to find.  This is my issue with the suggested replacement.  Now, while I'm not particularly thrilled with my own implementation, I do think it provides better transparency.

My current take is to leave it how it is, and change it when we finally decided on a standard to implement throughout the package.

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (namurphy) [inline plasmapy/analysis/swept_langmuir/ion_saturation_current.py:107]:
```suggestion
        Cannot be used with keyword ``voltage_bound``.
```

**Review by namurphy — COMMENTED** [review]
Looks good!  I had wondered about using the extract function refactoring pattern, but I don't think it'd make any real difference here.  I didn't look into the Langmuir analysis stuff, so after @svincena gives a review of that, I'd be happy to merge it soon after.  Thank you for doing this!

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (svincena) [inline plasmapy/analysis/swept_langmuir/ion_saturation_current.py:151]: becomes increasingly negative

**Review by svincena — APPROVED** [review]
Good to go, following our meeting today!

**Review by svincena — APPROVED** [review]

File: plasmapy/analysis/swept_langmuir/floating_potential.py
Comment (namurphy) [inline plasmapy/analysis/swept_langmuir/floating_potential.py:44]:
```suggestion
        1-D numpy array of monotonically ascending probe biases (must be in volts)

    current: `numpy.ndarray`
        1-D numpy array of probe current (must be in amperes) corresponding
```
[optional] If the function requires volts and amperes here, then we could be more specific with the language.  If not, feel free to ignore.

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (namurphy) [inline plasmapy/analysis/swept_langmuir/ion_saturation_current.py:63]:
```suggestion
```
[optional]

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (namurphy) [inline plasmapy/analysis/swept_langmuir/ion_saturation_current.py:69]:
```suggestion
        (must be in volts).

    current: `numpy.ndarray`
        1-D numpy array of probe current (must be in amperes)
```
Same optional comment as above.

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (namurphy) [inline plasmapy/analysis/swept_langmuir/ion_saturation_current.py:151]:
```suggestion
    becomes increasingly negative.  This sheath expansion results in the
```
This my only quasi-required comment, so if you fix this I can approve it.  All of my other suggestions are either optional or hilarious.

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (namurphy) [inline plasmapy/analysis/swept_langmuir/ion_saturation_current.py:150]: [optional] Perhaps break this up into two sentences?

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (namurphy) [inline plasmapy/analysis/swept_langmuir/ion_saturation_current.py:154]:
```suggestion
    rtn_extras = ISatExtras(rsq=None, fitted_func=None, fitted_indices=None)._asgard()
```
😬 ⚡ 🔨 

This was a really (t)horrible pun, if I do say so myself.

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (namurphy) [inline plasmapy/analysis/swept_langmuir/ion_saturation_current.py:176]:
```suggestion
            f"Requested fit {fit_type!r} is not a valid option.  Valid options "
```
[optional]

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (namurphy) [inline plasmapy/analysis/swept_langmuir/ion_saturation_current.py:200]: These variable names would also be appropriate in sea shanties.

> And in the mighty mighty storm 🎶 
> We sailed the current min. 🎵

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (namurphy) [inline plasmapy/analysis/swept_langmuir/ion_saturation_current.py:209]: I wondered if [masked arrays from NumPy](https://numpy.org/doc/stable/reference/maskedarray.html) would be useful here, but it doesn't seem like changing the implementation would improve maintainability much.

**Review by namurphy — COMMENTED** [review]
I have one quasi-required suggestion after which I'm happy to approve it.  All my other suggestions/comments are optional.

Thank you again for doing this, and I'm excited to be able to merge this soon!

**Review by namurphy — APPROVED** [review]
Actually, since we made the change to the settings today, I'll disable auto-merge and then approve it so that you can merge it when you're ready.  Thank you again!

File: plasmapy/analysis/swept_langmuir/floating_potential.py
Comment (rocco8773) [reply plasmapy/analysis/swept_langmuir/floating_potential.py:44]: I have to think about this. `voltage` and `current` should be in volts and amperes, respectively, but it is not needed for this function.  The Isat can be determined regardless of the units; it's just the returned Isat will share the same equivalent units as `voltage` and `current`.

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (rocco8773) [reply plasmapy/analysis/swept_langmuir/ion_saturation_current.py:154]: 🤦‍♂️

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (rocco8773) [reply plasmapy/analysis/swept_langmuir/ion_saturation_current.py:209]: I'll have to read more about masked arrays, but I don't think it's quite applicable here. Because, `mask` is being used to filter two independent arrays, `voltage` and `current`.  I'm also suspecting, without definitive proof, that my current approach is probably less resource heavy.

File: plasmapy/analysis/swept_langmuir/floating_potential.py
Comment (rocco8773) [reply plasmapy/analysis/swept_langmuir/floating_potential.py:44]: I wrote this thinking about $I_{sat}$, but the statement is still true for $V_f$.

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (rocco8773) [reply plasmapy/analysis/swept_langmuir/ion_saturation_current.py:175]: Interesting little trick.  Not sure how I feel about it.  As a coder I like that it could help with consistency, but I also feel like that comes with a penalty of less explicitness.  I don't think this is the original intent behind `!r` and there might be other behaviors hidden behind the veil.

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (namurphy) [reply plasmapy/analysis/swept_langmuir/ion_saturation_current.py:208]: Cool!  I'm perfectly content keeping it as is.

File: plasmapy/analysis/swept_langmuir/ion_saturation_current.py
Comment (StanczakDominik) [reply plasmapy/analysis/swept_langmuir/ion_saturation_current.py:179]: Oh wow, yeah, that's the first time I'm seeing it as well! Definitely a nice trick, but the previous implementation is way more obvious.

**Pull request merged**
